pkg/cluster/deploystorage.go

```diff
@@ -120,6 +154,112 @@ func (m *manager) ensureResourceGroup(ctx context.Context) (err error) {
 }
 
+func (m *manager) ensureTaggingPolicy(ctx context.Context) error {
+	displayName := tagPolicyDisplayName(m.doc.OpenShiftCluster.Properties.InfraID)
+	definition := resourceTaggingPolicyDefinition(displayName)
+	definition, err := m.definitions.CreateOrUpdate(ctx, displayName, definition)
+
+	if err != nil {
+		return err
+	}
+
+	parameters := map[string]*mgmtpolicy.ParameterValuesValue{}
+	i := 0
+
+	for k, v := range m.doc.OpenShiftCluster.Properties.ResourceTags {
+		tagKeyParamName := tagKeyParamName(i)
+		tagValueParamName := tagValueParamName(i)
+
+		parameters[tagKeyParamName] = &mgmtpolicy.ParameterValuesValue{
+			Value: k,
+		}
+
+		parameters[tagValueParamName] = &mgmtpolicy.ParameterValuesValue{
+			Value: v,
+		}
+
+		i++
+	}
+
+	// If the customer has passed fewer than the maximum possible number
+	// of tags, fill in the remaining parameters with empty strings.
+	for ; i < maxTags; i++ {
+		tagKeyParamName := tagKeyParamName(i)
+		tagValueParamName := tagValueParamName(i)
+
+		parameters[tagKeyParamName] = &mgmtpolicy.ParameterValuesValue{
+			Value: "",
+		}
+
+		parameters[tagValueParamName] = &mgmtpolicy.ParameterValuesValue{
+			Value: "",
+		}
+	}
+
+	assignment := mgmtpolicy.Assignment{}
+	assignment.Location = &m.doc.OpenShiftCluster.Location
+
+	assignment.AssignmentProperties = &mgmtpolicy.AssignmentProperties{
+		DisplayName:        &displayName,
+		PolicyDefinitionID: to.StringPtr(*definition.ID),
+		Scope:              to.StringPtr(m.doc.OpenShiftCluster.Properties.ClusterProfile.ResourceGroupID),
+		Parameters:         parameters,
+		EnforcementMode:    mgmtpolicy.Default,
+	}
+
+	assignment.Identity = &mgmtpolicy.Identity{
+		Type: mgmtpolicy.SystemAssigned,
+	}
+
+	assignment, err = m.assignments.Create(ctx, m.doc.OpenShiftCluster.Properties.ClusterProfile.ResourceGroupID, displayName, assignment)
+
+	if err != nil {
+		return err
+	}
+
+	roleAssignmentParams := mgmtauthorization.RoleAssignmentCreateParameters{
+		RoleAssignmentProperties: &mgmtauthorization.RoleAssignmentProperties{
+			RoleDefinitionID: to.StringPtr(tagContributorRoleDefinitionId),
+			PrincipalID:      assignment.Identity.PrincipalID,
+			PrincipalType:    mgmtauthorization.ServicePrincipal,
+		},
+	}
+
+	_, err = m.roleAssignments.Create(ctx, m.doc.OpenShiftCluster.Properties.ClusterProfile.ResourceGroupID, m.doc.OpenShiftCluster.Properties.UUID, roleAssignmentParams)
+
+	return err
+}
+
+func (m *manager) remediateTags(ctx context.Context) error {
+	resourceGroup := stringutils.LastTokenByte(m.doc.OpenShiftCluster.Properties.ClusterProfile.ResourceGroupID, '/')
+
+	resources, err := m.resources.ListByResourceGroup(ctx, resourceGroup, "", "", nil)
+
+	if err != nil {
+		return err
+	}
+
+	for _, r := range resources {
+		// Do not remove any tags, but add new tags and update values
+		// for existing tags.
+		for k, v := range m.doc.OpenShiftCluster.Properties.ResourceTags {
```

Comment:
One thought I had is that if the # resources in the cluster RG is greater than a certain amount, we could have `az aro update` error out and ask the customer to run a policy remediation task themselves. This would offload the work to Azure in those extreme cases.